### ISSUE TITLE
tcp_ipv6_v6only from libos/regression fails after kernel upgrade to 6.2 on linux native

### ISSUE THREAD

**Issue by anjalirai-intel** [gramineproject/gramine#1326]
### Description of the problem

We have started seeing tcp_ipv6_v6only failure after kernel upgrade from 6.0 to 6.2.

tcp_ipv6_v6only fails for linux native, outside of gramine environment.

It fails with error "bind(ipv4) was successful even though there is no IPV6_V6ONLY on same port".

we tried going through the v6.2 release notes and there were some changes related to ipv6 but not sure if its related.
**URL**: https://kernelnewbies.org/Linux_6.2
**Section:** Add IFF_NO_ADDRCONF to prevent ipv6 addrconf commit, commit, commit

I have attached the debug logs, if someone wants to debug this issue, please let me know we can share the machine details

### Steps to reproduce

Build libos/regression tests
Run tcp_ipv6_v6only test

### Expected results
**Debug Logs**
```
[P1:T1:tcp_ipv6_v6only] trace: ---- bind(3, {family=IPv6,ip=[0:0:0:0:0:0:0:0],port=11112}, 28) = 0x0
[P1:T1:tcp_ipv6_v6only] trace: ---- listen(3, 3) = 0x0
[P1:T1:tcp_ipv6_v6only] trace: ---- setsockopt(3, 1, 20, 0x22e6418e4d60, 16) = 0x0
[P1:T1:tcp_ipv6_v6only] trace: ---- setsockopt(3, 1, 21, 0x22e6418e4d60, 16) = 0x0
[P1:T1:tcp_ipv6_v6only] trace: ---- getsockopt(3, 1, 20, 0x22e6418e4d60, 0x22e6418e4d5c) = 0x0
[P1:T1:tcp_ipv6_v6only] trace: ---- getsockopt(3, 1, 21, 0x22e6418e4d60, 0x22e6418e4d5c) = 0x0
[P1:T1:tcp_ipv6_v6only] trace: ---- bind(4, {family=IPv4,ip=127.0.0.1,port=11112}, 16) = -98
```
test completed successfully

### Actual results
**Debug logs:** The bind operation should result in error -98 but it returned 0
```
(libos_parser.c:1606:buf_write_all) [P1:T1:tcp_ipv6_v6only] trace: ---- bind(3, {family=IPv6,ip=[0:0:0:0:0:0:0:0],port=11112}, 28) = 0x0
(libos_parser.c:1606:buf_write_all) [P1:T1:tcp_ipv6_v6only] trace: ---- listen(3, 3) = 0x0
(libos_parser.c:1606:buf_write_all) [P1:T1:tcp_ipv6_v6only] trace: ---- setsockopt(3, 1, 20, 0x1beaaf88ad70, 16) = 0x0
(libos_parser.c:1606:buf_write_all) [P1:T1:tcp_ipv6_v6only] trace: ---- setsockopt(3, 1, 21, 0x1beaaf88ad70, 16) = 0x0
(libos_parser.c:1606:buf_write_all) [P1:T1:tcp_ipv6_v6only] trace: ---- getsockopt(3, 1, 20, 0x1beaaf88ad70, 0x1beaaf88ad6c) = 0x0
(libos_parser.c:1606:buf_write_all) [P1:T1:tcp_ipv6_v6only] trace: ---- getsockopt(3, 1, 21, 0x1beaaf88ad70, 0x1beaaf88ad6c) = 0x0
(libos_parser.c:1606:buf_write_all) [P1:T1:tcp_ipv6_v6only] trace: ---- bind(4, {family=IPv4,ip=127.0.0.1,port=11112}, 16) = 0x0
```

bind(ipv4) was successful even though there is no IPV6_V6ONLY on same port"
[tcp_ipv6_v6only_199.log](https://github.com/gramineproject/gramine/files/11395466/tcp_ipv6_v6only_199.log)

### Gramine commit hash

babea62d7f927e1b9ba582f4f36f464eea82010b

**Comment by dimakuv** [comment]
The `IFF_NO_ADDRCONF` bug fixes in Linux 6.2 don't seem to be the root cause for your observed behavior. Someone needs to debug this. Could you also collect `strace gramine-direct ...` of both the good and the bad runs (i.e., what happens on the host)?

**Comment by kailun-qin** [comment]
I looked into this a bit. I suspect it's a regression caused by https://github.com/torvalds/linux/commit/28044fc1d4953b07acec0da4d2fc4784c57ea6fb (introduced in 6.1).

**Comment by anjalirai-intel** [comment]
@dimakuv I have attached the strace logs for both the scenarios

[strace_tcp_ipv6_v6only_199.log](https://github.com/gramineproject/gramine/files/11397585/strace_tcp_ipv6_v6only_199.log)
[strace_tcp_ipv6_v6only_110.log](https://github.com/gramineproject/gramine/files/11397588/strace_tcp_ipv6_v6only_110.log)

**Comment by mkow** [comment]
> ### Gramine commit hash
> 
> master

@anjalirai-intel: That's not a commit hash, please don't skip this field when filing issues.

**Comment by anjalirai-intel** [comment]
@mkow Actually, this issue is not because of Gramine ad commit hash is mandatory, that's why pointing to a specific commit does not made sense

**Comment by anjalirai-intel** [comment]
@kailun-qin @dimakuv Should we raise the issue with linux team?

**Comment by kailun-qin** [comment]
@anjalirai-intel Can we confirm first that the native failure was indeed caused by https://github.com/torvalds/linux/commit/28044fc1d4953b07acec0da4d2fc4784c57ea6fb? I made the speculation simply by kernel source/change inspection.

**Comment by anjalirai-intel** [comment]
@kailun-qin To validate this in native, we have to build kernel from sources, 

Is there any other way we can check this issue for native

**Comment by dimakuv** [comment]
> @kailun-qin To validate this in native, we have to build kernel from sources, Is there any other way we can check this issue for native.

I don't think there is any other way other than building the Linux kernel from sources.

But if you have contacts to the Linux team, then maybe someone from that team will be willing to check this? Because this looks like a real Linux bug/regression. A seasoned Linux developer will be able to check this very fast.

**Comment by mkow** [comment]
> @mkow Actually, this issue is not because of Gramine ad commit hash is mandatory, that's why pointing to a specific commit does not made sense

But the commit hash points to a specific version of Gramine and this very test. Saying "`master`" points nowhere, because in a year this may be a completely different code.

**Comment by kailun-qin** [comment]
> But the commit hash points to a specific version of Gramine and this very test. Saying "master" points nowhere, because in a year this may be a completely different code.

+1.

I reproduced the issue on my Linux v6.2 setup. I've updated the commit hash in the issue description (hope you don't mind @anjalirai-intel :)).

**Comment by kailun-qin** [comment]
FYI, @anjalirai-intel has reported the issue in the Linux regression report mailing list. The issue was confirmed (see https://lore.kernel.org/stable/20230508194406.73759-1-kuniyu@amazon.com/#t).

There was actually a same regression report (https://lore.kernel.org/netdev/e21bf153-80b0-9ec0-15ba-e04a4ad42c34@redhat.com/), and was fixed with this patch (https://lore.kernel.org/netdev/20230312031904.4674-2-kuniyu@amazon.com/). The patch was backported to v6.1.21 and v6.2.8, not v6.0.y (EOL).

To sum up, with Linux v6.3 (see https://github.com/torvalds/linux/blob/v6.3/net/ipv4/inet_hashtables.c), we should be able to see this issue gone.

**Comment by anjalirai-intel** [comment]
@kailun-qin I have verified the issue on v6.2.8 and the issue does not exist.

We can close the issue.

**Comment by dimakuv** [comment]
This is awesome! Thanks for hunting this down @anjalirai-intel and @kailun-qin.
